brave/src/main/java/brave/internal/baggage/BaggageFields.java

```diff
@@ -14,29 +14,25 @@
 package brave.internal.baggage;
 
 import brave.baggage.BaggageField;
-import brave.internal.Nullable;
-import brave.internal.Platform;
-import brave.internal.collect.LongBitSet;
 import brave.internal.collect.UnsafeArrayMap;
 import brave.internal.collect.UnsafeArrayMap.Mapper;
+import brave.internal.extra.MapExtra;
+import brave.internal.extra.MapExtraFactory;
 import brave.propagation.TraceContext;
 import brave.propagation.TraceContextOrSamplingFlags;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 
-import static brave.internal.baggage.BaggageFieldsHandler.MAX_DYNAMIC_FIELDS;
-import static brave.internal.collect.LongBitSet.isSet;
-import static brave.internal.collect.LongBitSet.setBit;
-
 /**
  * Holds one or more baggage fields in {@link TraceContext#extra()} or {@link
  * TraceContextOrSamplingFlags#extra()}.
  */
-public final class BaggageFields extends Extra<BaggageFields, BaggageFieldsFactory>
+public final class BaggageFields
+    extends MapExtra<BaggageField, String, BaggageFields, BaggageFields.Factory>
```

Comment:
you'll notice while the generics are a bit wild in the base type, the eventual result is fairly simple.